sources/model/Cloud.cpp

```diff
@@ -222,13 +371,23 @@ void Cloud::addGrain()
 {
     addFlag = true;
     myCloudVis->addGrain();
+    for (int i = 0; i < g_maxMidiVoices; i++)
+        if (playedCloudMidi[i]->isActive)
+            playedCloudMidi[i]->addFlag = true;
+        else
+            playedCloudMidi[i]->myGrains.push_back(new Grain(theSamples, duration, pitch));
     changed_numGrains = true;
 }
 
 void Cloud::removeGrain()
 {
     removeFlag = true;
     myCloudVis->removeGrain();
+    for (int i = 0; i < g_maxMidiVoices; i++)
+        if (playedCloudMidi[i]->isActive)
+            playedCloudMidi[i]->removeFlag = true;
+        else
+            playedCloudMidi[i]->myGrains.pop_back();
```

Comment:
Lorsque l'instance de `Grain` est retirée du tableau sans être `delete` au préalable, la mémoire fuite.
C'est vrai non seulement ici, mais aussi aux autre endroits qui appellent `pop_back` sur une telle liste. `Cloud` et `CloudMidi` sont tous deux concernés.
Par ailleurs, que ce passe-t-il si on entre ici alors que la liste de grains est vide ?

Reply:
ceci etait deja le cas , moi je me suis contenté de recopier ce que faisait cloud.
il suffit de rajouter un 
playedCloudMidi[i]->myGrains.delete avant le pop_back ?
et peut on avoir la liste de grains vide ? il me semble que le minimum est de 1 grain.

Reply:
> avant le pop_back ?

Oui, quelque chose comme ceci : `delete playedCloudMidi[i]->myGrains.back();`

>et peut on avoir la liste de grains vide ? il me semble que le minimum est de 1 grain.

Ok alors tout va bien.

Reply:
vu

Reply:
en fait , j'ai essaye 
`delete playedCloudMidi[i]->myGrains.back();`
à la place du pop_back, puis avant lui.
et j'ai un souci très compliqué avec un message de type 
`qt malloc(): smallbin double linked list corrupted` arrivant plus tard au bout de quelques boucles de jeu
apriori cette écriture est incorrecte.
je pense que ceci marche mieux : 
`delete playedCloudMidi[ii]->myGrains[playedCloudMidi[ii]->myGrains.size()-1];
                    playedCloudMidi[ii]->myGrains.pop_back();`
je ne semble plus avoir de problemes apres ça, mais j'avoue que je ne paitrise pas du tout c'est un peu au hasard.
qu'ne penses tu ?

Reply:
À propos de ces deux lignes, cela revient exactement à écrire
```
delete playedCloudMidi[ii]->myGrains.back();
playedCloudMidi[ii]->myGrains.pop_back();
```

Tu as un outil dans le compilateur permettant d'extraire une trace des erreurs mémoire ; poste ici éventuellement le rapport d'erreur, cela permettra d'y voir un peu clair.
https://blog.qt.io/blog/2013/04/17/using-gccs-4-8-0-address-sanitizer-with-qt/

Reply:
eh bien pourtant avec ce que j'ai mis ça ne plante plus alors qu'avec ta proposition si...

Reply:
C'est certain que non. J'ai lancé Frontières avec address sanitizer, et cela m'a signalé une erreur mémoire en relation avec les grains. (heap-buffer-overflow)
Le rapport d'erreur : https://gist.github.com/jpcima/fe48c13ffde951e6e189842f2906ec43

Reply:
Manifestement il pourrait s'agit d'un problème de concurrence : l'action `setNumGrain` est appelée depuis l'interface graphique, ce qui engendre la destruction des grains.

Par ailleurs, j'observe que le second site de destruction de grains se trouve dans la procédure audio. Lorsque le mutex de scène n'est pas verrouillé, la modification concurrente n'est pas sûre.

Note : la procédure audio ne devrait jamais appeler un `delete` out un `free`, ces opérations ne sont pas compatible-TR. C'est un élément du programme original qui reste à corriger.

Reply:
mais ce que tu me montres là n'est pas le code de la version qui est actuellement à jour dans le PR.
la version à jour fait ceci : 
`    for (int i = 0; i < g_maxMidiVoices; i++) {
        if (playedCloudMidi[i]->isActive){
            playedCloudMidi[i]->removeFlag = true;
        }
        else {
            delete playedCloudMidi[i]->myGrains[playedCloudMidi[i]->myGrains.size()-1];
            playedCloudMidi[i]->myGrains.pop_back();
        }`

Reply:
>mais j'ai bien mis le delete. ton passage à travers sanityser (auquel je ne comprends rien, ceci dit au passage) a ete fait sur quelle version ?

Sur ta version exacte des sources, à savoir 32f1d9572a135d2a02acc2c7107f6f9e78e2aef2 non modifiée, sans utiliser `.back()`. Observes tu un message d'erreur particulier ?

>sinon, la question de savoir où doivent être faites les destructions de grains, ça je ne sais pas trop pourquoi c'est fait à deux endroits.

Le plus simple (et sûr) consiste à faire depuis l'interface graphique toutes les constructions et destructions d'objets `Grain`, sous verrouillage du mutex de scène.

Il me semble que le code d'origine est clairement défaillant en ce qui concerne la gestion des grains. À mon avis, l'intention de départ du programmeur a été de permettre l'ajout et la suppression dynamique de grains sans interrompre la génération de son. (mais c'est un échec étant donné que `delete` n'est pas TR)

Dans un premier temps : je pense qu'il convient d'éliminer le fanion `removeFlag`, et tout `delete` au sein du fil audio. On permet à l'interface seule de faire tous suppressions et ajouts, sous possession du verrou de scène.
Dans un second temps, on peut élaborer une meilleure plus performante.

>il me faudrait connaitre tous les principes le la programmation audio TR pour pouvoir juger de ces questions, et j'apprends petit à petit surtout avec ce que tu m'en dis.
>
>existerait il une bonne documentation sur la question que je pourrais consulter ?

Or recommande assez souvent cet article qui fait assez bien le tour de la problématique TR. http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing

>quant au probleme de gain, peut etre rajouter un parametre gain dans les layers serait il la solution ?

À mon avis, mettre une formule de vélocité→volume qui n'atténue pas aussi brutalement dans le domaine des valeurs basses.

Reply:
> Sur ta version exacte des sources, à savoir [32f1d95](https://github.com/linuxmao-org/Frontieres/commit/32f1d9572a135d2a02acc2c7107f6f9e78e2aef2) non modifiée, sans utiliser `.back()`. Observes tu un message d'erreur particulier ?

euh, je ne comprends pas bien la question. me demandes tu si j'ai un message d'erreur telle quel actuellement sur la version 32f1d95 ? si c'est ça, la réponse est non, pas de messages particulier à la destruction de grain.

> Le plus simple (et sûr) consiste à faire depuis l'interface graphique toutes les constructions et destructions d'objets `Grain`, sous verrouillage du mutex de scène.
> 
> Il me semble que le code d'origine est clairement défaillant en ce qui concerne la gestion des grains. À mon avis, l'intention de départ du programmeur a été de permettre l'ajout et la suppression dynamique de grains sans interrompre la génération de son. (mais c'est un échec étant donné que `delete` n'est pas TR)
> 
> Dans un premier temps : je pense qu'il convient d'éliminer le fanion `removeFlag`, et tout `delete` au sein du fil audio. On permet à l'interface seule de faire tous suppressions et ajouts, sous possession du verrou de scène.
> Dans un second temps, on peut élaborer une solution plus performante.

je regarderai ces questions alors, je pense que ça devrait etre dans mpes cordes, et que ça va considérablement simplifier ma compreention du code en prime ! pour la solution plus performante, je m'en remets à toi...

> On recommande assez souvent cet article qui fait assez bien le tour de la problématique TR. http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing

merci, je me plonge dedans

> À mon avis, mettre une formule de vélocité→volume qui n'atténue pas aussi brutalement dans le domaine des valeurs basses.

je planche dessus